### PR TITLE
[16.0][MIG] account_invoice_margin: Migration to v16

### PR DESCRIPTION
**Pull request by ljsalvatierra-factorlibre**
https://github.com/OCA/margin-analysis/issues/142

### CODE REVIEW

**Comment by legalsylvain** [comment]
Hi. could you take a look on CI ? 
thanks !

**Comment by ljsalvatierra-factorlibre** [comment]
> Hi. could you take a look on CI ? thanks !

Hi, yes, i am resolving the issues.

**Comment by ljsalvatierra-factorlibre** [comment]
Hi, can anyone review this PR? Thanks in advance

**Comment by pedrobaeza** [comment]
You may find reviewers by chance if someone is interested, but a non selfish way of collaborating is to review other PRs, and ask in exchange that they review yours. When you are a good contributor, there will be other good contributors that will be will to make the reviews for free, but you have to equal the balance.

Anyway, if it's ready, remove the `[WIP]` label from the title.

**Comment by ljsalvatierra-factorlibre** [comment]
> You may find reviewers by chance if someone is interested, but a non selfish way of collaborating is to review other PRs, and ask in exchange that they review yours. When you are a good contributor, there will be other good contributors that will be will to make the reviews for free, but you have to equal the balance.
> 
> Anyway, if it's ready, remove the `[WIP]` label from the title.

Totally agree with you, i'll see what i can do to improve that :)

**Comment by legalsylvain** [comment]
Pre-commit is still red. Could you take a look ?

**Comment by ljsalvatierra-factorlibre** [comment]
> Pre-commit is still red. Could you take a look ?

Fixed, is it possible to re-run the tests with Odoo?

**Comment by legalsylvain** [comment]
if you rebase, it will rerun the CI.

**Comment by ljsalvatierra-factorlibre** [comment]
> if you rebase, it will rerun the CI.

Yes, but it is already up to date.

**Comment by ljsalvatierra-factorlibre** [comment]
Fixed one of the README images name.

File: account_invoice_margin/models/account_invoice.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        "invoice_line_ids.margin",
+        "invoice_line_ids.margin_signed",
+        "invoice_line_ids.price_subtotal",
+    )
+    def _compute_margin(self):
+        for invoice in self:
+            margin = 0.0
+            margin_signed = 0.0
+            price_subtotal = 0.0
+            for line in invoice._get_margin_applicable_lines():
```
Comment (legalsylvain) [inline account_invoice_margin/models/account_invoice.py:45]:
```suggestion
            for line in invoice.line_ids.filtered(lambda x: x.is_margin_applicable):
```

What about to have a compute field on account.invoice.line. I think that it's a caracteristic of the line.

Just a proposal, not a blocking point.

**Review by legalsylvain — COMMENTED** [review]
Hi. 

since OCA introduced _get_margin_applicable_lines features, margin can be computed correctly removing non relevant invoice lines. (down payment lines for example)

however, the sale_margin odoo module does'nt implement such feature and so the margin in sale.order can be wrong. (same for margin in pos.order) do you face the same issue ? 
If yes do you have an idea how to fix it ? 

thanks !

File: account_invoice_margin/models/account_invoice.py (hunk 1 above)
Comment (ljsalvatierra-factorlibre) [reply account_invoice_margin/models/account_invoice.py:45]: I'll take a look, thank you @legalsylvain for pointing it out.

Comment (legalsylvain) [reply account_invoice_margin/models/account_invoice.py:45]: Well, in fact, the ideal should be to have a uniq function _is_applicable_margin / _get_margin_applicable_lines or whatever that could be called in the account.move margin module / sale.order margin module / pos.order margin module.

But I'm not sure how to proceed.

**Comment by Yadier-Tecnativa** [comment]
ping @sergio-teruel and @pedrobaeza  TT42417

**Comment by gerard-vacas** [comment]
Hi.

To the comment that posted @legalsylvain I would comment:

This module is designed to calculate the margin in invoices and although in sale orders the margin could not be correct, we have that:
- Invoice will always recalculate margin from data in lines
- Even if using the module account_invoice_margin_sale, wich transfers the cost of the line from SO to invoice, and as far as I know, cost of down payments will be 0 at any case, margin will be calculated correctly in invoices

Based on that, I consider that any fixes on sale order margins would be great but I doubt this module is the one for them.

Having said that, the module LGTM

Thanks

**Review by JavierIniesta — APPROVED** [review]

**Comment by rafaelbn** [comment]
Runbot not available to test, please @ljsalvatierra-factorlibre could you make a rebase or add an empty commit trying to wake up runboat? 😄

**Comment by ljsalvatierra-factorlibre** [comment]
> Runbot not available to test, please @ljsalvatierra-factorlibre could you make a rebase or add an empty commit trying to wake up runboat? smile

Done, thank you for the reminder! :)

File: account_invoice_margin/models/account_invoice.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        compute="_compute_purchase_price",
+        store=True,
+        readonly=False,
+        digits="Product Price",
+    )
+
+    @api.depends("purchase_price", "price_subtotal")
+    def _compute_margin(self):
+        for line in self:
+            if line.move_id and line.move_id.move_type[:2] == "in":
```
Comment (RodrigoBM) [inline account_invoice_margin/models/account_invoice.py]:
```suggestion
            if line.move_id and line.move_id.is_purchase_document():
```

File: account_invoice_margin/models/account_invoice.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+                line.update(
+                    {"margin": 0.0, "margin_signed": 0.0, "margin_percent": 0.0}
+                )
+                continue
+            tmp_margin = line.price_subtotal - (line.purchase_price * line.quantity)
+            sign = line.move_id.move_type in ["in_refund", "out_refund"] and -1 or 1
```
Comment (RodrigoBM) [inline account_invoice_margin/models/account_invoice.py]:
```suggestion
            sign = 1 if self.move_id.is_inbound() else -1```

File: account_invoice_margin/models/account_invoice.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    def _get_purchase_price(self):
+        # Overwrite this function if you don't want to base your
+        # purchase price on the product standard_price
+        self.ensure_one()
+        return self.product_id.standard_price
+
+    @api.depends("product_id", "product_uom_id")
+    def _compute_purchase_price(self):
+        for line in self:
+            if line.move_id.move_type in ["out_invoice", "out_refund"]:
```
Comment (RodrigoBM) [inline account_invoice_margin/models/account_invoice.py]:
```suggestion
            if line.move_id.move_type.is_sale_document():
```

File: account_invoice_margin/tests/test_account_invoice_margin.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+from odoo.tests.common import Form, tagged
+
+from odoo.addons.account.tests.common import AccountTestInvoicingCommon
+
+
+@tagged("post_install", "-at_install")
+class TestAccountInvoiceMargin(AccountTestInvoicingCommon):
+    @classmethod
+    def setUpClass(cls):
+        super(TestAccountInvoiceMargin, cls).setUpClass()
```
Comment (RodrigoBM) [inline account_invoice_margin/tests/test_account_invoice_margin.py]: Apply PEP 3135 (28-Apr-2007) to "super()."
```suggestion
        super().setUpClass()
```

**Review by RodrigoBM — CHANGES_REQUESTED** [review]

**Review by Gelo-fl — APPROVED** [review]
Functional revision OK

**Review by RodrigoBM — APPROVED** [review]

**Review by JavierIniesta — APPROVED** [review]

**Comment by RodrigoBM** [comment]
Is it possible to merge this PR? @OCA/accounting-maintainers

**Comment by sergio-teruel** [comment]
ping @juangomezzurita

**Pull request merged**